File: ecommerce-frontend/src/features/shop/services/walletService.ts
import api from '@/lib/axios';

export interface Wallet {
  id: string;
  userId: string;
  walletType: string;
  availableBalance: number;
  escrowBalance: number;
  createdAt: string;
  updatedAt: string;
}

export interface WalletTransaction {
  id: string;
  walletId: string;
  orderId: string | null;
  amount: number;
  transactionType: string;
  description: string;
  createdAt: string;
}

export interface WithdrawalRequest {
  id: string;
  sellerId: string;
  amount: number;
  bankName: string;
  bankAccountNumber: string;
  bankAccountName: string;
  status: string; // PENDING, APPROVED, REJECTED
  createdAt: string;
  updatedAt: string;
}

export interface WalletDetailsResponse {
  wallet: Wallet;
  transactions: WalletTransaction[];
  withdrawals: WithdrawalRequest[];
}

export const walletService = {
  getMyWalletDetails: () =>
    api.get<WalletDetailsResponse>('/api/users/v1/wallets/me').then((r) => r.data),

  requestWithdrawal: (data: {
    amount: number;
    bankName: string;
    bankAccountNumber: string;
    bankAccountName: string;
  }) =>
    api.post<{ message: string }>('/api/users/v1/wallets/withdraw', data).then((r) => r.data),

  getAdminWithdrawals: () =>
    api.get<WithdrawalRequest[]>('/api/users/v1/wallets/admin/withdrawals').then((r) => r.data),

  processWithdrawal: (requestId: string, isApproved: boolean) =>
    api.post<{ message: string }>(`/api/users/v1/wallets/admin/withdrawals/${requestId}/process`, {
      isApproved,
    }).then((r) => r.data),
};

File: ecommerce-backend/user-service/src/main/java/com/ecommerce/user_service/service/WalletService.java
package com.ecommerce.user_service.service;

import com.ecommerce.user_service.entity.Wallet;
import com.ecommerce.user_service.entity.WalletTransaction;
import com.ecommerce.user_service.entity.WithdrawalRequest;
import java.util.List;

public interface WalletService {
    void createWalletIfNotExists(String userId, String walletType);
    void addAvailableBalance(String userId, Double amount, String description, String orderId);
    void subtractAvailableBalance(String userId, Double amount, String description, String orderId);
    void addEscrowBalance(String userId, Double amount, String description, String orderId);
    void subtractEscrowBalance(String userId, Double amount, String description, String orderId);
    void releaseEscrowToAvailable(String userId, Double amount, String description, String orderId);
    void requestWithdrawal(String sellerId, Double amount, String bankName, String bankAccountNumber, String bankAccountName);
    void processWithdrawal(String requestId, boolean isApproved);

    Wallet getWalletByUserId(String userId);
    List<WalletTransaction> getTransactionsByWalletId(String walletId);
    List<WithdrawalRequest> getWithdrawalRequestsBySellerId(String sellerId);
    List<WithdrawalRequest> getAllWithdrawalRequests();
}
